### PR TITLE
Improve documentation of Run() and RunManager()

### PR DESCRIPTION
**Pull request by heplesser**
The critical point of this PR is that in the `RunManager` example, the time argument is missing to the `Run()` call. I also added more information on calls one must not use between `prepare` and `cleanup`.

### CODE REVIEW

File: pynest/nest/lib/hl_api_simulation.py [hunk 1]
```diff
@@ -84,7 +83,10 @@ def Run(t):
     `Prepare` must be called before `Run` to calibrate the system, and
     `Cleanup` must be called after `Run` to close files, cleanup handles, and
     so on. After `Cleanup`, `Prepare` can and must be called before more `Run`
-    calls. Any calls to `SetStatus` between `Prepare` and `Cleanup` have
+    calls.
+
+    Any calls modifying the network, i.e., calls to `Create`, `Connect`,
+    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
     undefined behaviour.
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:90]:
```suggestion
    undefined behavior.
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 2]
```diff
@@ -139,8 +141,16 @@ def RunManager():
...
+            for _ in range(10):
+                Run(100)
+                # extract results
+
+    Notes
+    -----
+
+    Any calls modifying the network, i.e., calls to `Create`, `Connect`,
+    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
+    undefined behaviour.
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:153]:
```suggestion
    undefined behavior.
```

**Review by sarakonradi — CHANGES_REQUESTED** [review]
Thanks! Just added a minor comment regarding the American spelling.

File: pynest/nest/lib/hl_api_simulation.py (hunk 1 above)
Comment (tfardet) [inline pynest/nest/lib/hl_api_simulation.py:90]: ``Disconnect`` is not mentioned there and I have honestly no idea whether it'd be safe to call it or not...

Comment (heplesser) [reply pynest/nest/lib/hl_api_simulation.py]: I added it!

**Comment by heplesser** [comment]
@sarakonradi I fixed the spelling in both places.

**Review by sarakonradi — APPROVED** [review]

**Comment by tfardet** [comment]
I was just contacted by a student from my old team working with NEST and using ``Disconnect``. Apparently he managed to consistently use ``Disconnect`` within the ``RunManager`` context so I'd like to postpone this until tomorrow when he should bring up an issue with some example code to check and discuss this.
I'm not sure whether it works because of an overlooked issue or if we can actually get properly defined behavior there, but since there should be not handshake and whatnot required to disconnect nodes, I guess it may actually be possible to use it safely.
Until we're sure about that, I'd rather put the addition of ``Disconnect`` in the doc on hold (especially in case we need to specify that using it is actually fine).

**Comment by heplesser** [comment]
@sarakonradi @tfardet I just revised the notes on modifying between prepare and cleanup, could you re-review?

File: pynest/nest/lib/hl_api_simulation.py [hunk 3]
```diff
@@ -85,9 +85,13 @@ def Run(t):
     so on. After `Cleanup`, `Prepare` can and must be called before more `Run`
     calls.
 
-    Any calls modifying the network, i.e., calls to `Create`, `Connect`, `Disconnect`,
-    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
-    undefined behavior.
+    Be careful about modifying the network or neurons between `Prepare` and `Cleanup`
+    calls, especially do not `Create` and `Connect` neurons or call `SetKernelStatus`.
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:89]:
``` suggestion
    calls, especially not to `Create` and `Connect` neurons or call `SetKernelStatus`.
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 4]
```diff
@@ -148,9 +152,12 @@ def RunManager():
     Notes
     -----
 
-    Any calls modifying the network, i.e., calls to `Create`, `Connect`,
-    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
-    undefined behavior.
+    Be careful about modifying the network or neurons inside the `RunManager` context,
+    especially do not `Create` and `Connect` neurons or call `SetKernelStatus`. Calling
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:155]:
``` suggestion
    especially not to `Create` and `Connect` neurons or call `SetKernelStatus`. Calling
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 5]
```diff
@@ -85,9 +85,13 @@ def Run(t):
     so on. After `Cleanup`, `Prepare` can and must be called before more `Run`
     calls.
 
-    Any calls modifying the network, i.e., calls to `Create`, `Connect`, `Disconnect`,
-    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
-    undefined behavior.
+    Be careful about modifying the network or neurons between `Prepare` and `Cleanup`
+    calls, especially do not `Create` and `Connect` neurons or call `SetKernelStatus`.
+    Calling `SetStatus` to change membrane potential `V_m` of neurons or change synaptic
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:90]:
``` suggestion
    Calling `SetStatus` to change membrane potential `V_m` of neurons or synaptic
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 6]
```diff
@@ -148,9 +152,12 @@ def RunManager():
     Notes
     -----
 
-    Any calls modifying the network, i.e., calls to `Create`, `Connect`,
-    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
-    undefined behavior.
+    Be careful about modifying the network or neurons inside the `RunManager` context,
+    especially do not `Create` and `Connect` neurons or call `SetKernelStatus`. Calling
+    `SetStatus` to change membrane potential `V_m` of neurons or change synaptic weights
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:156]:
``` suggestion
    `SetStatus` to change membrane potential `V_m` of neurons or synaptic weights
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 7]
```diff
@@ -85,9 +85,13 @@ def Run(t):
...
 
-    Any calls modifying the network, i.e., calls to `Create`, `Connect`, `Disconnect`,
-    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
-    undefined behavior.
+    Be careful about modifying the network or neurons between `Prepare` and `Cleanup`
+    calls, especially do not `Create` and `Connect` neurons or call `SetKernelStatus`.
+    Calling `SetStatus` to change membrane potential `V_m` of neurons or change synaptic
+    weights (but not delays!) will in most cases work as expected, while changing
+    membrane or synaptic times constants **will not work correctly**. If in doubt, assume
+    that changes may cause undefined behavior and check very carefully.
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:93]:
``` suggestion
    that changes may cause undefined behavior and check these thoroughly.
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 8]
```diff
@@ -148,9 +152,12 @@ def RunManager():
...
 
-    Any calls modifying the network, i.e., calls to `Create`, `Connect`,
-    `SetStatus` and `SetKernelStatus` between `Prepare` and `Cleanup` have
-    undefined behavior.
+    Be careful about modifying the network or neurons inside the `RunManager` context,
+    especially do not `Create` and `Connect` neurons or call `SetKernelStatus`. Calling
+    `SetStatus` to change membrane potential `V_m` of neurons or change synaptic weights
+    (but not delays!) will in most cases work as expected, while changing membrane or
+    synaptic times constants **will not work correctly**. If in doubt, assume that changes
+    may cause undefined behavior and check very carefully.
```
Comment (sarakonradi) [inline pynest/nest/lib/hl_api_simulation.py:159]:
``` suggestion
    may cause undefined behavior and check these thoroughly.
```

**Review by sarakonradi — CHANGES_REQUESTED** [review]
@heplesser Thanks for revising the notes. I have a few minor suggestions.

**Comment by heplesser** [comment]
@sarakonradi Thanks, all accepted!

**Review by sarakonradi — APPROVED** [review]
Looks good to me!

File: pynest/nest/lib/hl_api_simulation.py [hunk 9]
```diff
@@ -84,8 +83,14 @@ def Run(t):
     `Prepare` must be called before `Run` to calibrate the system, and
     `Cleanup` must be called after `Run` to close files, cleanup handles, and
     so on. After `Cleanup`, `Prepare` can and must be called before more `Run`
-    calls. Any calls to `SetStatus` between `Prepare` and `Cleanup` have
-    undefined behaviour.
+    calls.
+
+    Be careful about modifying the network or neurons between `Prepare` and `Cleanup`
+    calls, especially not to `Create` and `Connect` neurons or call `SetKernelStatus`.
```
Comment (tfardet) [inline pynest/nest/lib/hl_api_simulation.py:89]:
```suggestion
    calls. In particular, do not call `Create`, `Connect`, or `SetKernelStatus`.
```

File: pynest/nest/lib/hl_api_simulation.py [hunk 10]
```diff
@@ -139,8 +144,19 @@ def RunManager():
...
+
+    Notes
+    -----
+
+    Be careful about modifying the network or neurons inside the `RunManager` context,
+    especially not to `Create` and `Connect` neurons or call `SetKernelStatus`. Calling
+    `SetStatus` to change membrane potential `V_m` of neurons or synaptic weights
+    (but not delays!) will in most cases work as expected, while changing membrane or
+    synaptic times constants **will not work correctly**. If in doubt, assume that changes
+    may cause undefined behavior and check these thoroughly.
```
Comment (tfardet) [inline pynest/nest/lib/hl_api_simulation.py:159]:
```suggestion
    ```suggestion
    Be careful about modifying the network or neurons inside the `RunManager` context.
    In particular, do not call `Create`, `Connect`, or `SetKernelStatus`. Calling `SetStatus`
    to change membrane potential `V_m` of neurons or synaptic weights (but not delays!)
    will in most cases work as expected, while changing membrane or synaptic times
    constants **will not work correctly**. If in doubt, assume that changes may cause
    undefined behavior and check these thoroughly.
```

**Review by tfardet — CHANGES_REQUESTED** [review]
Hi, sorry for the long delay.
I would propose to modify the sentence slightly to make it clear that ``Create``, ``Connect`` and ``SetKernelStatus`` are a hard no, it did not feel explicit enough to me with the previous sentence since it sounded that you should just be careful when calling them.

**Comment by heplesser** [comment]
@tfardet Thank you for your suggestions, I have accepted them all.

File: pynest/nest/lib/hl_api_simulation.py [hunk 11]
```diff
@@ -139,8 +144,20 @@ def RunManager():
...
-            for i in range(10):
-                Run()
+            for _ in range(10):
+                Run(100)
+                # extract results
+
+    Notes
+    -----
+
+    ```suggestion
```
Comment (tfardet) [inline pynest/nest/lib/hl_api_simulation.py:154]: remove

**Review by tfardet — CHANGES_REQUESTED** [review]
A "suggestion" crept in.
I also wonder whether the "do not" on the second sentence should be bold to match the "will not work correctly"... your call...

File: pynest/nest/lib/hl_api_simulation.py (hunk 11 above)
Comment (heplesser) [reply pynest/nest/lib/hl_api_simulation.py]: @tfardet Thanks for catching this, fixed. I decided to remove the bold everywhere, one tends to overuse it ;).

File: pynest/nest/lib/hl_api_simulation.py [hunk 12]
```diff
@@ -84,8 +83,14 @@ def Run(t):
...
     so on. After `Cleanup`, `Prepare` can and must be called before more `Run`
-    calls. Any calls to `SetStatus` between `Prepare` and `Cleanup` have
-    undefined behaviour.
+    calls.
+
+    Be careful about modifying the network or neurons between `Prepare` and `Cleanup`
+    calls. In particular, do not call `Create`, `Connect`, or `SetKernelStatus`.
+    Calling `SetStatus` to change membrane potential `V_m` of neurons or synaptic
+    weights (but not delays!) will in most cases work as expected, while changing
+    membrane or synaptic times constants **will not work correctly**. If in doubt, assume
```
Comment (tfardet) [inline pynest/nest/lib/hl_api_simulation.py:92]: I thought you wanted to remove the bold part (I don't mind though)

**Review by tfardet — APPROVED** [review]
I think maybe you forgot one bold section but I'll let you decide whether you want to remove it or not.
I'm fine with the current state.

File: pynest/nest/lib/hl_api_simulation.py (hunk 12 above)
Comment (heplesser) [reply pynest/nest/lib/hl_api_simulation.py]: Thanks for catching this, dangerous to have essentially the same comment in two places ;). Fixed now.

**Pull request merged**